_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,694 @@
+            "Validation error: is your website a reachable URL?",
+        ));
+    }
+
+    sqlx::query("INSERT INTO websites (url, alias) VALUES ($1, $2)")
+        .bind(new_website.url)
+        .bind(new_website.alias)
+        .execute(&state.db)
+        .await
+        .unwrap();
+
+    Ok(Redirect::to("/"))
+```
+
+Now let’s write a route to grab all the websites we’re monitoring, as well as get a quick report on what the uptime was like in the last recorded 24 hours (filling in any gaps where required) for each website. 
+
+Like before with the monitoring tasks, we need to grab a list of all of the websites we’re currently tracking and add them to a vector of website data (except we will assume there are results there. If not, `askama` will handle it for us by automatically not rendering any records):
+
+```rust
+#[derive(Serialize, Validate)]
+struct WebsiteInfo {
+    #[validate(url)]
+    url: String,
+    alias: String,
+    data: Vec<WebsiteStats>,
+}
+
+#[derive(sqlx::FromRow, Serialize)]
+pub struct WebsiteStats {
+    time: DateTime<Utc>,
+    uptime_pct: Option<i16>,
+}
+
+async fn get_websites(State(state): State<AppState>) -> Result<impl AskamaIntoResponse, ApiError> {
+    let websites = sqlx::query_as::<_, Website>("SELECT url, alias FROM websites")
+        .fetch_all(&state.db)
+        .await?;
+
+    let mut logs = Vec::new();
+
+    for website in websites {
+        let data = get_daily_stats(&website.alias, &state.db).await?;
+
+        logs.push(WebsiteInfo {
+            url: website.url,
+            alias: website.alias,
+            data,
+        });
+    }
+
+    Ok(WebsiteLogs { logs })
+```
+
+Once this is done, we’ll then need to make a new Vector. This will hold all the website URLs (and respective aliases) as well as a list of timestamps with the uptime percentage over the last 24 hours, calculated per hour from how many HTTP requests returned with `200 OK`. We then need to check if there’s any gaps and fill them in with a `None`. This lets the person viewing the data know that no data was recorded at the time (for example if we’re either developing locally, or if the service itself had an outage and was unable to record data). 
+
+To start with, let’s write a function for getting the daily stats for a given URL:
+
+```rust
+#[derive(sqlx::FromRow, Serialize)]
+pub struct WebsiteStats {
+    time: DateTime<Utc>,
+    uptime_pct: Option<i16>,
+}
+
+async fn get_daily_stats(alias: &str, db: &PgPool) -> Result<Vec<WebsiteStats>, ApiError> {
+    let data = sqlx
+        ::query_as::<_, WebsiteStats>(
+            r#"
+            SELECT date_trunc('hour', created_at) as time,
+            CAST(COUNT(case when status = 200 then 1 end) * 100 / COUNT(*) AS int2) as uptime_pct
+            FROM logs
+            LEFT JOIN websites on websites.id = logs.website_id
+            WHERE websites.alias = $1
+            group by time
+            order by time asc
+            limit 24
+            "#
```

Comment:
This query still has a logical bug in that it can return 24 rows that span more than 24 hours (between the first row and last row). Again, imagine what this will return if the service was down for 3 hours about 5 hours ago - the missing 3 hours will come from 25, 26 and 27 hours ago.